File: exercises/bestTimeToBuy121.go
package main

// import "fmt"

func maxProfit(prices []int) int {
    buyPrice := 100000 
    profit := 0
    for _, price := range prices {
        if (price - buyPrice) > profit {
            profit = price - buyPrice
        }
        if price < buyPrice {
            buyPrice = price
        }
    }
    return profit
}

File: exercises/best_time_to_buy_121.java

public class bestTimeToBuy121 {
    public static void main(String[] args) {
        // var sol = new Solution();
        // var res = sol.maxProfit(new int[] {7, 5, 4, 2, 1});
        // out.printf("The result is %d %n", res);
    }
}

class Solution_121 {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int buyPrice = Integer.MAX_VALUE;
        for (int price : prices) {
            profit = (price - buyPrice) > profit ? (price - buyPrice) : profit;
            buyPrice = price < buyPrice ? price : buyPrice;
        }
        return profit;
    }
}

